ksp_plugin/geometric_potential_plotter.cpp

```diff
@@ -0,0 +1,213 @@
+    not_null<Ephemeris<Barycentric>*> ephemeris)
+    : ephemeris_(ephemeris) {}
+
+void GeometricPotentialPlotter::Interrupt() {
+  plotter_ = jthread();
+  // We are single-threaded here, no need to lock.
+  plotter_idle_ = true;
+}
+
+void GeometricPotentialPlotter::RequestEquipotentials(
+    Parameters const& parameters) {
+  last_parameters_ = parameters;
+  absl::MutexLock l(&lock_);
+  // Only process this request if there is no analysis in progress.
+  if (plotter_idle_) {
+    plotter_idle_ = false;
+    plotter_ = MakeStoppableThread(
+        [this, parameters]() { PlotEquipotentials(parameters).IgnoreError(); });
+  }
+}
+
+std::optional<GeometricPotentialPlotter::Parameters> const&
+GeometricPotentialPlotter::last_parameters()
+    const {
+  return last_parameters_;
+}
+
+void GeometricPotentialPlotter::RefreshEquipotentials() {
+  absl::MutexLock l(&lock_);
+  if (next_equipotentials_.has_value()) {
+    equipotentials_ = std::move(next_equipotentials_);
+    next_equipotentials_.reset();
+  }
+}
+
+GeometricPotentialPlotter::Equipotentials const*
+GeometricPotentialPlotter::equipotentials() const {
+  return equipotentials_.has_value() ? &*equipotentials_ : nullptr;
+}
+
+absl::Status GeometricPotentialPlotter::PlotEquipotentials(
+    Parameters const& parameters) {
+  Instant const& t = parameters.time;
+
+  auto const& primary =
+      parameters.primary->mass() > parameters.secondary->mass()
+          ? *parameters.primary
+          : *parameters.secondary;
+  auto const& secondary =
+      parameters.primary->mass() > parameters.secondary->mass()
+          ? *parameters.secondary
+          : *parameters.primary;
+
+  RotatingPulsatingReferenceFrame<Barycentric, Navigation> const
+      reference_frame(ephemeris_, parameters.primary, parameters.secondary);
+  auto const plane =
+      Plane<Navigation>::OrthogonalTo(Vector<double, Navigation>({0, 0, 1}));
+
+  auto const potential = [&reference_frame,
+                          &t](Position<Navigation> const& position) {
+    return reference_frame.GeometricPotential(t, position);
+  };
+  auto const acceleration = [&reference_frame,
+                             &t](Position<Navigation> const& position) {
+    auto const acceleration = reference_frame.GeometricAcceleration(
+        t, {position, Velocity<Navigation>{}});
+    // Note the sign.
+    return -Vector<Acceleration, Navigation>({acceleration.coordinates()[0],
+                                              acceleration.coordinates()[1],
+                                              Acceleration{}});
+  };
+
+  const MultiLevelSingleLinkage<SpecificEnergy, Position<Navigation>, 2>::Box
+      box = {.centre = Navigation::origin,
+             .vertices = {
+                 Displacement<Navigation>({3 * Metre, 0 * Metre, 0 * Metre}),
+                 Displacement<Navigation>({0 * Metre, 3 * Metre, 0 * Metre})}};
+
+  constexpr Length characteristic_length = 1 * Nano(Metre);
```

Comment:
Is it so small because the box is 3 metres long?  A comment would help.  Or maybe a constant that is the "box size", and everything scaled according to it.  See also the computation of wells, below.